transaction/std/src/memo_builder/gift_code_cancellation_memo_builder.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Defines the Memo Builder for the gift code cancellation memo (0x0202)
+//! specified in MCIP #32
+
+use super::{
+    memo::{GiftCodeCancellationMemo, UnusedMemo},
+    MemoBuilder, ReservedDestination,
+};
+use mc_account_keys::PublicAddress;
+use mc_transaction_core::{Amount, MemoContext, MemoPayload, NewMemoError};
+
+/// There are three possible gift code memo types specified in MCIP #32
+/// | Memo type bytes | Name                        |
+/// | -----------     | -----------                 |
+/// |    0x0002       | Gift Code Sender Memo       |
+/// |    0x0201       | Gift Code Funding Memo      |
+/// | -->0x0202<--    | Gift Code Cancellation Memo |
+/// This memo builder builds a gift code cancellation memo (0x0202). Gift code
+/// cancellation is defined as the sender sending the gift code TxOut at the
+/// gift code subaddress back to their default address prior to it being spent
+/// by the receiver. When that happens a change TxOut is sent to the gift
+/// code sender's change subaddress with a gift code cancellation memo that
+/// stores the index of the TxOut originally sent to the gift code subaddress
+/// when the gift code was funded.
+#[derive(Clone, Debug)]
+pub struct GiftCodeCancellationMemoBuilder {
+    // Index of the gift code TxOut that was originally funded
+    gift_code_tx_out_global_index: Option<u64>,
```

Comment:
I mainly did it to make it consistent a consistent flow with the other builders providing opt-outs of certain flows. But I agree in general with memo builders being "all-or-nothing" and doing the thing they are designed to do and using the unused memo builder if you're not planning to use it.

Reply:
In other words, I'm going remove the optional change output from all the gift code memo builders.

Reply:
FWIW I agree with Eran, if you don't have this number then the reader won't be able to make sense of what gift code was cancelled. So it should be required -- IMO we should not implement `Default`, we should require the user of this class to provide the number in a `new` function

This is similar to what we did in the `BurnRedemptionMemoBuilder`.

In the `RTHMemoBuilder` we didn't require the sender memo credential, but that was only because we thought there might be a use-case where a user chooses not to use the sender memo thing, but still wants change outputs. (Partly this was from discussions with Koe.) And so when that data was omitted it writes unused memo. If we didn't think that behavior was desired then probably `RTHMemoBuilder` would require a sender memo credential to be constructed.

In this case though, I think we can't actually use the memo correctly at all if we don't have this number.